### ISSUE TITLE
Components: Improve Toolbar

### ISSUE THREAD

**Issue by diegohaz** [WordPress/gutenberg#18619]
I'm creating this issue to place together the steps necessary to improve both the API and the accessibility of the `Toolbar` component in `@wordpress/components`. This work started on #17875, but, since the PR has become too complex, [we've chosen to break it into more parts](https://github.com/WordPress/gutenberg/pull/17875#issuecomment-553866039):

- [x] [**#18534**] Introduce an `__experimental` accessible `Toolbar` component in `@wordpress/components`

- [x] [**#18931**] Improve `ToolbarButton` so that it serves more use cases than just being an `IconButton`.

- [x] [**#19242**] Refactor `SlotFill` so we won't [need to force fills to re-render when `fillProps` change](https://github.com/WordPress/gutenberg/pull/17875#discussion_r342428408).

- [x] [**#20008**] Convert the fixed toolbar buttons (block switcher, more tools etc.) to `ToolbarButton`.

- [x] [**#22354**] Convert header toolbar.

- [x] [**#22696**] [**#23216**] [**#23252**] [**#23277**] [**#23278**] [**#23281**] Convert all core block toolbars.

    This will fix #15331 and #3383.

- [ ] [**#23316**] If it proves to be useful and works out, deprecate the old usage and remove the `__experimental` prefix. Otherwise, revert it altogether, which would break the code for people using `__experimentalAccessibilityLabel` explicitly, but custom blocks would seamlessly rollback to the old usage.

**Comment by gziolo** [comment]
Re-sharing my comment https://github.com/WordPress/gutenberg/pull/18534#issuecomment-555413744:

> How about we also move `packages/components/src/toolbar-context/index.js` to be located inside `packages/components/src/toolbar/context.js`? Well, it might wait until we decide whether we need to expose it. My only concern is that it gives the impression that this is something that could use standalone since it is in the root folder.
> 
> There is also `README.md` missing for the newly introduced `ToolbarGroup` component. Given that JSDoc comments contain all required defaults, it should be quite straightforward to create a documentation file. A separate story is that it would be nice to be able to auto-generate such files based on JSDoc and Storybook stories :D

**Comment by gziolo** [comment]
> This will fix #15331 and #3383.

Can we close those issues? It seems like everything is ready. The only task left is deprecation warnings for the old approach, right?

**Comment by diegohaz** [comment]
> Can we close those issues? It seems like everything is ready. The only task left is deprecation warnings for the old approach, right?

There are still some blocks missing. I'll make a list of them.

**Comment by diegohaz** [comment]
Missing blocks:

- [ ] Table (after adding rows/columns)
- [ ] Navigation (vertical and horizontal, with content)
- [ ] Custom HTML
- [ ] All embeds (Edit URL button)
